sim-se: Implement free-list-based physical page allocator for SE mode

nmosier opened the pull request:
This series of commits implements a free-list-based physical page allocator for SE mode, supported by the following:
* 700a5a5 Adds a new stat, `maxAllocatedBytes`, to memory pools, which tracks the maximum physical memory, in bytes, that has been allocated at any point in the simulation so far.
* 6cfa219 Fixes an issue with `statistics::Group` that made gem5 crash after making `MemPool` a stats group
* 385fc6e Implements the free-list physical page allocator itself.

These patches fix issue #1809.

I evaluated the old, naive physical page allocator ("old" in the table) versus the new, free-list-based physical page allocator ("new" in the table) on the SPEC CPU2017 Integer benchmarks, using the X86KvmCPU model in SE mode. I collected three stats:
* **sim mem**: the physical memory allocated in the simulator (given in the `maxAllocatedBytes` stats), labeled as "sim mem" in the table
* **host mem**: the maximum memory usage of the gem5 process (collected via `/usr/bin/time -v`)
* **host time**: the host runtime to run the benchmark to completion.
<table>
    <tr>
        <th rowspan="2">benchmark</th>
        <th colspan="2">sim mem (maxAllocatedBytes, in GiB)</th>
        <th colspan="2">host mem (GiB)</th>
        <th colspan="2">host runtime (s)</th>
    </tr>
    <tr>
        <th>old</th><th>new</th>
        <th>old</th><th>new</th>
        <th>old</th><th>new</th>
    </tr>
    <tr><td>perlbench.0</td><td>0.20</td><td>0.20</td><td>0.28</td><td>0.29</td><td>74.91</td><td>75.43</td></tr>
    <tr><td>perlbench.1</td><td>0.16</td><td>0.15</td><td>0.24</td><td>0.23</td><td>43.86</td><td>43.88</td></tr>
    <tr><td>perlbench.2</td><td>0.09</td><td>0.09</td><td>0.17</td><td>0.17</td><td>69.68</td><td>71.45</td></tr>
    <tr><td>gcc.0</td><td>9.29</td><td>7.48</td><td>9.47</td><td>7.67</td><td>344.63</td><td>391.42</td></tr>
    <tr><td>gcc.1</td><td>3.24</td><td>3.05</td><td>3.36</td><td>3.18</td><td>101.30</td><td>100.75</td></tr>
    <tr><td>gcc.2</td><td>3.29</td><td>3.04</td><td>3.42</td><td>3.17</td><td>99.43</td><td>100.97</td></tr>
    <tr><td>mcf</td><td>13.22</td><td>4.83</td><td>12.40</td><td>4.96</td><td>429.66</td><td>477.23</td></tr>
    <tr><td>omnetpp</td><td>0.27</td><td>0.24</td><td>0.36</td><td>0.33</td><td>214.43</td><td>214.22</td></tr>
    <tr><td>xalancbmk</td><td>0.47</td><td>0.47</td><td>0.58</td><td>0.58</td><td>129.34</td><td>129.84</td></tr>
    <tr><td>x264.0</td><td>0.15</td><td>0.15</td><td>0.23</td><td>0.23</td><td>16.95</td><td>17.10</td></tr>
    <tr><td>x264.1</td><td>0.04</td><td>0.04</td><td>0.12</td><td>0.12</td><td>44.75</td><td>44.60</td></tr>
    <tr><td>x264.2</td><td>0.15</td><td>0.15</td><td>0.23</td><td>0.23</td><td>46.70</td><td>46.64</td></tr>
    <tr><td>deepsjeng</td><td>6.73</td><td>6.73</td><td>6.89</td><td>6.89</td><td>250.32</td><td>250.02</td></tr>
    <tr><td>leela</td><td>0.10</td><td>0.03</td><td>0.18</td><td>0.10</td><td>237.70</td><td>237.85</td></tr>
    <tr><td>exchange2</td><td>0.00</td><td>0.00</td><td>0.08</td><td>0.08</td><td>319.68</td><td>320.10</td></tr>
    <tr><td>xz.0</td><td>26.06</td><td>15.03</td><td>26.54</td><td>15.51</td><td>592.75</td><td>656.06</td></tr>
    <tr><td>xz.1</td><td>6.15</td><td>4.11</td><td>6.51</td><td>4.48</td><td>1173.43</td><td>1182.33</td></tr>
    <tr><td>total</td><td>69.61</td><td>45.80</td><td>71.06</td><td>48.22</td><td>4189.52</td><td>4359.89</td></tr>
    <tr><td>total, normalized</td><td>1</td><td>0.66</td><td>1</td><td>0.68</td><td>1</td><td>1.04</td></tr>
</table>

The results show that using the new free-list-based physical page allocator significantly reduces the amount of RAM required to run the SPEC benchmarks in gem5 in SE mode. There is a 4% increase in runtime overhead. This is simply due to the need to zero out physical pages upon deallocation. Notably, the runtime is only significantly slower for benchmarks that free a lot of memory (e.g., xz.0 and xz.1) and thus have significantly reduced memory requirements — that is, we pay for reducing memory overhead with a slight increase in runtime overhead. However, this runtime overhead becomes negligible for simulated CPUs (rather than KVM-based CPUs).

nmosier committed:
sim-se: implement free-list page allocator

Fix #1824. Implement a free-list-based physical page allocator.
The previous allocator was naive and never reallocated pages
that were freed (e.g., by the `munmap` system call). This resulted
simulated workloads requiring in some cases much larger memories
than when running natively on the host (in gcc SPEC2017 benchmark,
up to 4x more).

This commit implements replaces the naive page allocator with
slightly less naive one that uses a simple free list to track
ranges of physical pages that are free. The free list is included
in checkpoints.

Now, whenever a virtual address is unmapped (which, with this patch,
now only occurs in `MemState::unmapRegion`), its mapped-to physical
page is marked as deallocated/freed and thus available to be
reallocated at some later point.

Change-Id: I51c691cf18372ae298f5743feefd6b0ca20a581c

giactra commented:
Hi @nmosier; this looks like a pretty useful change.

I have two initial thoughts

1) Can't we make page zeroing an optional behaviour? In most cases we are more concerned about simulation speed rather than modelling the kernel zeroing behaviour (which happens at allocation anyway and not deallocation)

2) Could you elaborate a bit further on the stats change? You say the problem happens when we destruct a stats group. But this shouldn't really happen as we might lose important information. Is it a problem now because we deallocate MemPool? Wouldn't it make more sense to add the stats to MemPools class instead?

erin-le commented:
@nmosier Hello, thank you for your contribution! When you have time, could you respond to the comments above?

halfgodxzc commented:
> @nmosier Hello, thank you for your contribution! When you have time, could you respond to the comments above?

@erin-le  Hi, I’ve reviewed the commit thoroughly and am very interested in the changes. This is my first time participating in an issue, and I wanted to ask if it would be acceptable for me to contribute by writing code to address the concerns mentioned in the comments.

erin-le commented:
@halfgodxzc It doesn't look like the author is actively working on this PR, so please go ahead!

nmosier commented:
@giactra Thanks for taking a look and the feedback. I addressed your comments and submitted a revised patch.
1. I made page zeroing optional by adding the `zeroPages` param to the `Process` python object. Generally, zeroing pages is necessary because glibc malloc assumes that all newly mmap'ed pages are zero-filled. See issue #342 and pull request #343 for more details.
2. I chose to zero pages out upon deallocation rather than allocation to avoid unnecessarily zeroing out pages that have never been allocated before and thus are already zero-filled. I added a comment at src/sim/process.cc:362 explaining this.
3. I moved the `maxAllocatedBytes` stats from `MemPool` to `MemPools` as you suggested, and was thus able to avoid mucking around with stats group destructors. The new patch does not modify anything in src/base now.

powerjg commented on src/base/free_list.cc:
Why is this empty file needed? Shouldn't the code in `free_list.hh` be compiled when it is included in `process.cc`? Also, it's not compiled in the `SConscript`.

nmosier replied:
Ah, not sure why I added that. I removed it in the latest patch.

powerjg commented on src/sim/Process.py:
```diff
@@ -51,6 +51,10 @@ def map(self, vaddr, paddr, size, cacheable=False):
     maxStackSize = Param.MemorySize("64MiB", "maximum size of the stack")
+    zeroPages = Param.Bool(
+        True,
+        "ensure all allocated pages are zero-filled. glibc malloc generally requires this. Disable at your own risk.",
```
This line is beyond 80 characters

nmosier replied:
Oops, fixed!

powerjg commented on src/sim/mem_pool.cc:
```diff
@@ -114,31 +95,70 @@ MemPool::totalBytes() const
+        freePhysPages.insert(free_page_num, _totalPages - free_page_num);
+    } else {
+        ScopedCheckpointSection sec(cp, "free_list");
+        int n;
```
```suggestion
        int n = 0;
```

Just to make sure the static analysis tools won't report this as an uninitialized variable.

nmosier replied:
Fixed!

powerjg reviewed (commented):
Thanks for the updates. Overall, this looks good to me. I have a couple of small things below.

nmosier commented:
Thanks! I just addressed your three comments with my latest patch.

powerjg commented:
@giactra can you take a look at this after the response from @nmosier

giactra commented on src/base/free_list.test.cc:
```diff
@@ -0,0 +1,96 @@
+
+using namespace gem5;
+
+TEST(FreeListTest, Simple)
```
Adding these unit tests is great but I feel they are not stressing some tricky corner cases, like for example what happens if you start inserting overlapping ranges

Judging from https://github.com/gem5/gem5/pull/1825/files#diff-544445642128ae52856649383a6912f556c5de8ea74af7400893c82c4b6b6b3bR74 I have the feeling this is not covered by your FreeList implementation

nmosier replied:
Correct, the free list doesn't support inserting overlapping ranges, i.e., it doesn't permit double frees. I added a few extra test cases to verify that the free list panics when a double free is attempted.

giactra commented on src/base/free_list.hh:
```diff
@@ -0,0 +1,159 @@
+     * Allocate a region of size @param size out of free space.
+     * @return whether the allocation succeeded.
+     */
+    bool
+    allocate(T size, T& base)
```
Nit, how do you feel about:

```
std::optional<T>
allocate(T size)
{
```

In this way you don't have to pass the return value by reference

nmosier replied:
Sure, makes sense. I made that change.

giactra commented on src/sim/mem_pool.cc:
```diff
@@ -148,10 +168,30 @@ MemPools::populate(const AddrRangeList &memories)
-    return pools[pool_id].allocate(npages);
+    const Addr addr = pools[pool_id].allocate(npages);
+
+    // Recompute max allocated bytes stat, since it may have changed.
+    stats.maxAllocatedBytes =
+        std::max<Addr>(stats.maxAllocatedBytes.value(), allocatedBytes());
```
I think this needs a bit of attention

When dealing with stats one always needs to take into consideration they can (and will) be reset manually with the m5 reset pseudo op before entering a ROI.

So at a certain point maxAllocatedBytes might be reset to 0.

What is the stat supposed to model? Is it supposed to model the maximum number of allocated bytes in the entire simulation (m5 reset shouldn't affect it)? Or do we want to model the number of allocated bytes in a specific time window (m5 reset should affect it)

powerjg replied:
> m5 reset shouldn't affect it

Let's take this discussion off of this thread, but I'm not sure we should have any "stats" which are not reset. We've had confusion in the past with this. Maybe we should have a different word/interface for stats which don't reset.

See #1615 and #1699 for examples.

Let's discuss this at a high level at the next developers meeting.

nmosier replied:
I removed this stat to simplify the patch, since it's not necessary for the free list to work properly.

giactra reviewed (commented):
Thanks @nmosier for your changes! Please check my comments below

erin-le commented:
Hello @nmosier, when you have time, could you address these comments? Thank you!

erin-le commented:
@nmosier Hello, just another reminder about the comments on this PR. Thank you for your contribution!

nmosier commented:
@giactra I pushed a new version that tries to address your feedback. Apologies for the delay.

giactra reviewed (approved):
Thanks @nmosier for your changes!
